[Doc] Add doc about offline cache (#5646)
taichi-dev/taichi: docs/lang/articles/performance/performance.md

As a rule of thumb, run benchmarks to decide whether to enable BLS or not.
:::

## Offline Cache

> A taichi kernel is implicitly compiled the first time it is invoked. The compilation yieldings are kept in an "online" in-memory cache to reduce the overhead in subsequent function calls. As long as the kernel function is unchanged, it can be directly loaded and launched. However, the cache is no longer availabe when the entire program runs down to its termination. Taichi re-compiles all the kernel functions and constructs the "online" in-memory cache when the user runs the program again. The first launch of a Taichi function is always slow due to compilation overhead.

Reviewer: A Taichi kernel is implicitly compiled the first time it is called. The compilation results are kept in an *online* in-memory cache to reduce the overhead in the subsequent function calls. As long as the kernel function is unchanged, it can be directly loaded and launched. The cache, however, is no longer available when the program terminates. Then, if you run the program again, Taichi has to re-compile all kernel functions and reconstruct the *online* in-memory cache. And the first launch of a Taichi function is always slow due to the compilation overhead.

Reviewer: Could we just omit "in-memory"?

Author: IMO, maybe "in-memory" can underline the online cache that Taichi had is on the memory level, but the new feature, "offline cache," is on the filesystem level? WDYT?

> We address the problem by introducing the "offline" cache feature, that could dump and save the compilation cache on disk for future runs. The first launch overhead can be drastically reduced in repeated runs. Taichi now automatically constructs and maintains the offline cache by default. We also provide several options in `ti.init` to configure the detailed behaviors of offline cache.

Reviewer: ```suggestion
We address this problem by introducing the *offline* cache feature, which dumps and saves the compilation cache on disk for future runs. The first launch overhead can be drastically reduced in repeated runs. Taichi now constructs and maintains an offline cache by default, as well as providing several options in `ti.init()` for configuring the offline cache behavior.
```

Reviewer: I'm not quite certain about "The first launch overhead can be drastically reduced in repeated runs.". Could I ask you to elaborate on this one? I thought the first launch overhead would stay unchanged and would not be reduced?

Author: The overhead of "the first launch" includes compilation and the truly 1st run. If the launched Taichi kernel's compilation data was saved on disk, the next "the first launch" overhead would be reduced because of **offline cache**.

> * `offline_cache: bool`: Enable or disable offline cache. Default: `True`.

Reviewer: ```suggestion
* `offline_cache: bool`: Enables or disables offline cache. Default: `True`.
```

* `offline_cache_file_path: str`: The offline cache file directory. Default: `'C:\taichi_cache\ticache\'` on Windows and `'~/.cache/taichi/ticache/'` on unix-like systems. Directories are created automatically.
* `offline_cache_max_size_of_files: int32`: Max size of cache files in bytes. Trigger cleaning process when the cache file size exceeds the given capacity. Default: 100MB.

> * `offline_cache_cleaning_policy: str`: The policy used to replace out-of-date files in the cache. Optional polices: `'never'`, `'version'`, `'lru'` or `'fifo'`. Default: `'lru'`.

Reviewer: ```suggestion
* `offline_cache_cleaning_policy: str`: Policy about how to replace outdated files in the cache. Options: `'never'`, `'version'`, `'lru'` and `'fifo'`. Default: `'lru'`.
```

>   * `'never'`: Never clean, keep all the cache files. Ignore the `offline_cache_max_size_of_files` configuration;

Reviewer: ```suggestion
  * `'never'`: Never cleans and keeps all the cached files regardless of the `offline_cache_max_size_of_files` configuration;
```

>   * `'version'`: Only discard old-version cache files with respect to the kernel function;

Reviewer: ```suggestion
  * `'version'`: Discards only the old-version cached files with respect to the kernel function;
```

>   * `'lru'`: Discard the cache files that are least recently used;

Reviewer: ```suggestion
  * `'lru'`: Discards the cached files least used recently;
```

>   * `'fifo'`: Discard the cache files that are earliest added.

Reviewer: ```suggestion
  * `'fifo'`: Discards the cached files added in the earliest.
```
